mb2ff.1
.Dd 2024-03-26
.Dt MB2FF 1
.Os farbfeld-utils
.Sh NAME
.Nm mb2ff
.Nd convert Hero Mesh puzzle image to farbfeld
.Sh SYNOPSIS
.Nm
<letter> <?class?> <?index?>
.Sh DESCRIPTION
.Nm
reads a Free Hero Mesh composite puzzle set from stdin, extracts pictures and converts them to
.Xr farbfeld 5
and writes the result to stdout.
.Pp
Free Hero Mesh is a free/open-source puzzle game engine where you can define your own classes of
objects using the Free Hero Mesh programming language.
.Pp
A Free Hero Mesh composite puzzle set file is a Hamster archive containing the
files of a Free Hero Mesh puzzle set, and optionally some additional lumps,
such as FILE_ID.DIZ.
.Pp
Free Hero Mesh doesn't care about the file name of a composite puzzle set; the use of .fhm is merely
a convention used by some people.
.Pp
In case of an error
.Nm
writes a diagnostic message to stderr.
.Sh EXIT STATUS
.Bl -tag -width Ds
.It 0
Image processed successfully.
.It 1
An error occurred.
.El
.Sh Options
This decoder takes up to three arguments:
   letter -- a single letter, from A to C, which specifies what size of
             pictures are wanted (this is the only mandatory argument)
   class  -- a class name or a one-based level number. If a class, export all
             pictures in a vertical strip. If a level number, render that
             level.
   index  -- a zero-based image number within the class. This is only valid if
             the second argument is a class name. It outputs only that one
             picture, rather than all of them.

Pictures are sixteen per row (use ff-strip to change this) and are in the
allocation order; not the order they appear in class definitions.

Note when previewing a level, it renders it as it would be in the editor; at
runtime it may look different due to MSG_INIT codes altering stuff.

Also see:
   - http://fileformats.archiveteam.org/wiki/Free_Hero_Mesh_composite_puzzle_set
   - http://zzo38computer.org/fossil/heromesh.ui/index
   - https://chiselapp.com/user/zzo38/repository/freeheromesh/info/27aef9b
.Sh EXAMPLES
$
.Nm
A < image.fhm > image.ff
.Pp
$
.Nm
B 3 < image.fhm > image.ff
.Sh SEE ALSO
.Xr fhm2ff 1 ,
.Xr farbfeld 5 ,
.Xr farbfeld-utils 7


mb2ff.c
#if 0
gcc -s -O2 -o ~/bin/mb2ff -Wno-unused-result mb2ff.c
exit
#endif

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

static const unsigned char palette[0x800]={
0x00,0x00,0x00,0x00,0x00,0x00,0x00,0x00, // 0xC0,0x20,0xFF,
0x00,0x00,0x00,0x00,0x00,0x00,0xFF,0xFF, 0x22,0x22,0x22,0x22,0x22,0x22,0xFF,0xFF, 0x33,0x33,0x33,0x33,0x33,0x33,0xFF,0xFF, 0x44,0x44,0x44,0x44,0x44,0x44,0xFF,0xFF,
0x55,0x55,0x55,0x55,0x55,0x55,0xFF,0xFF, 0x66,0x66,0x66,0x66,0x66,0x66,0xFF,0xFF, 0x77,0x77,0x77,0x77,0x77,0x77,0xFF,0xFF, 0x88,0x88,0x88,0x88,0x88,0x88,0xFF,0xFF,
0x99,0x99,0x99,0x99,0x99,0x99,0xFF,0xFF, 0xAA,0xAA,0xAA,0xAA,0xAA,0xAA,0xFF,0xFF, 0xBB,0xBB,0xBB,0xBB,0xBB,0xBB,0xFF,0xFF, 0xCC,0xCC,0xCC,0xCC,0xCC,0xCC,0xFF,0xFF,
0xDD,0xDD,0xDD,0xDD,0xDD,0xDD,0xFF,0xFF, 0xEE,0xEE,0xEE,0xEE,0xEE,0xEE,0xFF,0xFF, 0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF, 0x28,0x28,0x14,0x14,0x00,0x00,0xFF,0xFF,
0x41,0x41,0x23,0x23,0x00,0x00,0xFF,0xFF, 0x5F,0x5F,0x32,0x32,0x00,0x00,0xFF,0xFF, 0x84,0x84,0x21,0x21,0x00,0x00,0xFF,0xFF, 0xA0,0xA0,0x50,0x50,0x00,0x00,0xFF,0xFF,
0xC3,0xC3,0x5F,0x5F,0x14,0x14,0xFF,0xFF, 0xE1,0xE1,0x73,0x73,0x1E,0x1E,0xFF,0xFF, 0xFF,0xFF,0x82,0x82,0x32,0x32,0xFF,0xFF, 0xFF,0xFF,0x91,0x91,0x41,0x41,0xFF,0xFF,
0xFF,0xFF,0xA0,0xA0,0x50,0x50,0xFF,0xFF, 0xFF,0xFF,0xAF,0xAF,0x5F,0x5F,0xFF,0xFF, 0xFF,0xFF,0xBE,0xBE,0x73,0x73,0xFF,0xFF, 0xFF,0xFF,0xD2,0xD2,0x82,0x82,0xFF,0xFF,
0xFF,0xFF,0xE1,0xE1,0x91,0x91,0xFF,0xFF, 0xFF,0xFF,0xF0,0xF0,0xA0,0xA0,0xFF,0xFF, 0x32,0x32,0x1E,0x1E,0x1E,0x1E,0xFF,0xFF, 0x41,0x41,0x22,0x22,0x20,0x20,0xFF,0xFF,
0x5F,0x5F,0x28,0x28,0x30,0x30,0xFF,0xFF, 0x82,0x82,0x30,0x30,0x40,0x40,0xFF,0xFF, 0xA0,0xA0,0x3A,0x3A,0x4C,0x4C,0xFF,0xFF, 0xBE,0xBE,0x46,0x46,0x58,0x58,0xFF,0xFF,
0xE1,0xE1,0x54,0x54,0x64,0x64,0xFF,0xFF, 0xFF,0xFF,0x66,0x66,0x70,0x70,0xFF,0xFF, 0xFF,0xFF,0x7F,0x7F,0x7B,0x7B,0xFF,0xFF, 0xFF,0xFF,0x8E,0x8E,0x7F,0x7F,0xFF,0xFF,
0xFF,0xFF,0x9F,0x9F,0x7F,0x7F,0xFF,0xFF, 0xFF,0xFF,0xAF,0xAF,0x7F,0x7F,0xFF,0xFF, 0xFF,0xFF,0xBF,0xBF,0x7F,0x7F,0xFF,0xFF, 0xFF,0xFF,0xCF,0xCF,0x7F,0x7F,0xFF,0xFF,
0xFF,0xFF,0xDF,0xDF,0x7F,0x7F,0xFF,0xFF, 0x28,0x28,0x0D,0x0D,0x0D,0x0D,0xFF,0xFF, 0x40,0x40,0x15,0x15,0x15,0x15,0xFF,0xFF, 0x60,0x60,0x20,0x20,0x20,0x20,0xFF,0xFF,
0x80,0x80,0x2A,0x2A,0x2A,0x2A,0xFF,0xFF, 0xA0,0xA0,0x35,0x35,0x35,0x35,0xFF,0xFF, 0xC0,0xC0,0x40,0x40,0x40,0x40,0xFF,0xFF, 0xE0,0xE0,0x4A,0x4A,0x4A,0x4A,0xFF,0xFF,
0xFF,0xFF,0x55,0x55,0x55,0x55,0xFF,0xFF, 0xFF,0xFF,0x67,0x67,0x64,0x64,0xFF,0xFF, 0xFF,0xFF,0x6F,0x6F,0x64,0x64,0xFF,0xFF, 0xFF,0xFF,0x75,0x75,0x84,0x84,0xFF,0xFF,
0xFF,0xFF,0x84,0x84,0x9D,0x9D,0xFF,0xFF, 0xFF,0xFF,0x94,0x94,0xB7,0xB7,0xFF,0xFF, 0xFF,0xFF,0x9F,0x9F,0xD1,0xD1,0xFF,0xFF, 0xFF,0xFF,0xAE,0xAE,0xEA,0xEA,0xFF,0xFF,
0x90,0x90,0x14,0x14,0x00,0x00,0xFF,0xFF, 0xA0,0xA0,0x20,0x20,0x00,0x00,0xFF,0xFF, 0xB0,0xB0,0x30,0x30,0x00,0x00,0xFF,0xFF, 0xC0,0xC0,0x40,0x40,0x00,0x00,0xFF,0xFF,
0xD0,0xD0,0x50,0x50,0x00,0x00,0xFF,0xFF, 0xE0,0xE0,0x60,0x60,0x00,0x00,0xFF,0xFF, 0xF0,0xF0,0x70,0x70,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0x80,0x80,0x00,0x00,0xFF,0xFF,
0xFF,0xFF,0x90,0x90,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0xA0,0xA0,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0xB0,0xB0,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0xC0,0xC0,0x00,0x00,0xFF,0xFF,
0xFF,0xFF,0xD0,0xD0,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0xE0,0xE0,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0xF0,0xF0,0x00,0x00,0xFF,0xFF, 0x28,0x28,0x00,0x00,0x00,0x00,0xFF,0xFF,
0x40,0x40,0x00,0x00,0x00,0x00,0xFF,0xFF, 0x60,0x60,0x00,0x00,0x00,0x00,0xFF,0xFF, 0x80,0x80,0x00,0x00,0x00,0x00,0xFF,0xFF, 0xA0,0xA0,0x00,0x00,0x00,0x00,0xFF,0xFF,
0xC0,0xC0,0x00,0x00,0x00,0x00,0xFF,0xFF, 0xE0,0xE0,0x00,0x00,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0x00,0x00,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0x28,0x28,0x28,0x28,0xFF,0xFF,
0xFF,0xFF,0x40,0x40,0x40,0x40,0xFF,0xFF, 0xFF,0xFF,0x60,0x60,0x60,0x60,0xFF,0xFF, 0xFF,0xFF,0x80,0x80,0x80,0x80,0xFF,0xFF, 0xFF,0xFF,0xA0,0xA0,0xA0,0xA0,0xFF,0xFF,
0xFF,0xFF,0xC0,0xC0,0xC0,0xC0,0xFF,0xFF, 0xFF,0xFF,0xE0,0xE0,0xE0,0xE0,0xFF,0xFF, 0x28,0x28,0x00,0x00,0x28,0x28,0xFF,0xFF, 0x40,0x40,0x00,0x00,0x40,0x40,0xFF,0xFF,
0x60,0x60,0x00,0x00,0x60,0x60,0xFF,0xFF, 0x80,0x80,0x00,0x00,0x80,0x80,0xFF,0xFF, 0xA0,0xA0,0x00,0x00,0xA0,0xA0,0xFF,0xFF, 0xC0,0xC0,0x00,0x00,0xC0,0xC0,0xFF,0xFF,
0xE0,0xE0,0x00,0x00,0xE0,0xE0,0xFF,0xFF, 0xFF,0xFF,0x00,0x00,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0x28,0x28,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0x40,0x40,0xFF,0xFF,0xFF,0xFF,
0xFF,0xFF,0x60,0x60,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0x80,0x80,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0xA0,0xA0,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0xC0,0xC0,0xFF,0xFF,0xFF,0xFF,
0xFF,0xFF,0xE0,0xE0,0xFF,0xFF,0xFF,0xFF, 0x28,0x28,0x14,0x14,0x28,0x28,0xFF,0xFF, 0x40,0x40,0x20,0x20,0x40,0x40,0xFF,0xFF, 0x60,0x60,0x30,0x30,0x60,0x60,0xFF,0xFF,
0x80,0x80,0x40,0x40,0x80,0x80,0xFF,0xFF, 0xA0,0xA0,0x50,0x50,0xA0,0xA0,0xFF,0xFF, 0xC0,0xC0,0x60,0x60,0xC0,0xC0,0xFF,0xFF, 0xE0,0xE0,0x70,0x70,0xE0,0xE0,0xFF,0xFF,
0xFF,0xFF,0x7C,0x7C,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0x8C,0x8C,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0x9C,0x9C,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0xAC,0xAC,0xFF,0xFF,0xFF,0xFF,
0xFF,0xFF,0xBC,0xBC,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0xCC,0xCC,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0xDC,0xDC,0xFF,0xFF,0xFF,0xFF, 0xFF,0xFF,0xEC,0xEC,0xFF,0xFF,0xFF,0xFF,
0x28,0x28,0x00,0x00,0x50,0x50,0xFF,0xFF, 0x35,0x35,0x05,0x05,0x66,0x66,0xFF,0xFF, 0x42,0x42,0x0A,0x0A,0x7C,0x7C,0xFF,0xFF, 0x4F,0x4F,0x0F,0x0F,0x92,0x92,0xFF,0xFF,
0x5C,0x5C,0x14,0x14,0xA8,0xA8,0xFF,0xFF, 0x69,0x69,0x19,0x19,0xBE,0xBE,0xFF,0xFF, 0x76,0x76,0x1E,0x1E,0xD4,0xD4,0xFF,0xFF, 0x83,0x83,0x23,0x23,0xEA,0xEA,0xFF,0xFF,
0x90,0x90,0x28,0x28,0xFF,0xFF,0xFF,0xFF, 0xA0,0xA0,0x40,0x40,0xFF,0xFF,0xFF,0xFF, 0xB0,0xB0,0x60,0x60,0xFF,0xFF,0xFF,0xFF, 0xC0,0xC0,0x80,0x80,0xFF,0xFF,0xFF,0xFF,
0xD0,0xD0,0xA0,0xA0,0xFF,0xFF,0xFF,0xFF, 0xE0,0xE0,0xC0,0xC0,0xFF,0xFF,0xFF,0xFF, 0xF0,0xF0,0xE0,0xE0,0xFF,0xFF,0xFF,0xFF, 0x00,0x00,0x00,0x00,0x28,0x28,0xFF,0xFF,
0x00,0x00,0x00,0x00,0x40,0x40,0xFF,0xFF, 0x00,0x00,0x00,0x00,0x60,0x60,0xFF,0xFF, 0x00,0x00,0x00,0x00,0x80,0x80,0xFF,0xFF, 0x00,0x00,0x00,0x00,0xA0,0xA0,0xFF,0xFF,
0x00,0x00,0x00,0x00,0xC0,0xC0,0xFF,0xFF, 0x00,0x00,0x00,0x00,0xE0,0xE0,0xFF,0xFF, 0x00,0x00,0x00,0x00,0xFF,0xFF,0xFF,0xFF, 0x0A,0x0A,0x28,0x28,0xFF,0xFF,0xFF,0xFF,
0x28,0x28,0x4A,0x4A,0xFF,0xFF,0xFF,0xFF, 0x46,0x46,0x6A,0x6A,0xFF,0xFF,0xFF,0xFF, 0x67,0x67,0x8A,0x8A,0xFF,0xFF,0xFF,0xFF, 0x87,0x87,0xAA,0xAA,0xFF,0xFF,0xFF,0xFF,
0xA7,0xA7,0xCA,0xCA,0xFF,0xFF,0xFF,0xFF, 0xC7,0xC7,0xEB,0xEB,0xFF,0xFF,0xFF,0xFF, 0x0F,0x0F,0x1E,0x1E,0x1E,0x1E,0xFF,0xFF, 0x14,0x14,0x23,0x23,0x23,0x23,0xFF,0xFF,
0x19,0x19,0x32,0x32,0x32,0x32,0xFF,0xFF, 0x1E,0x1E,0x41,0x41,0x41,0x41,0xFF,0xFF, 0x28,0x28,0x50,0x50,0x50,0x50,0xFF,0xFF, 0x32,0x32,0x5F,0x5F,0x5F,0x5F,0xFF,0xFF,
0x37,0x37,0x73,0x73,0x73,0x73,0xFF,0xFF, 0x41,0x41,0x82,0x82,0x82,0x82,0xFF,0xFF, 0x46,0x46,0x91,0x91,0x91,0x91,0xFF,0xFF, 0x50,0x50,0xA0,0xA0,0xA0,0xA0,0xFF,0xFF,
0x5A,0x5A,0xAF,0xAF,0xAF,0xAF,0xFF,0xFF, 0x5F,0x5F,0xC3,0xC3,0xC3,0xC3,0xFF,0xFF, 0x69,0x69,0xD2,0xD2,0xD2,0xD2,0xFF,0xFF, 0x73,0x73,0xE1,0xE1,0xE1,0xE1,0xFF,0xFF,
0x78,0x78,0xF0,0xF0,0xF0,0xF0,0xFF,0xFF, 0x00,0x00,0x28,0x28,0x28,0x28,0xFF,0xFF, 0x00,0x00,0x40,0x40,0x40,0x40,0xFF,0xFF, 0x00,0x00,0x60,0x60,0x60,0x60,0xFF,0xFF,
0x00,0x00,0x80,0x80,0x80,0x80,0xFF,0xFF, 0x00,0x00,0xA0,0xA0,0xA0,0xA0,0xFF,0xFF, 0x00,0x00,0xC0,0xC0,0xC0,0xC0,0xFF,0xFF, 0x00,0x00,0xE0,0xE0,0xE0,0xE0,0xFF,0xFF,
0x00,0x00,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF, 0x28,0x28,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF, 0x40,0x40,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF, 0x60,0x60,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
0x80,0x80,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF, 0xA0,0xA0,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF, 0xC0,0xC0,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF, 0xE0,0xE0,0xFF,0xFF,0xFF,0xFF,0xFF,0xFF,
0x00,0x00,0x28,0x28,0x00,0x00,0xFF,0xFF, 0x00,0x00,0x40,0x40,0x00,0x00,0xFF,0xFF, 0x00,0x00,0x60,0x60,0x00,0x00,0xFF,0xFF, 0x00,0x00,0x80,0x80,0x00,0x00,0xFF,0xFF,
0x00,0x00,0xA0,0xA0,0x00,0x00,0xFF,0xFF, 0x00,0x00,0xC0,0xC0,0x00,0x00,0xFF,0xFF, 0x00,0x00,0xE0,0xE0,0x00,0x00,0xFF,0xFF, 0x00,0x00,0xFF,0xFF,0x00,0x00,0xFF,0xFF,
0x28,0x28,0xFF,0xFF,0x28,0x28,0xFF,0xFF, 0x40,0x40,0xFF,0xFF,0x40,0x40,0xFF,0xFF, 0x60,0x60,0xFF,0xFF,0x60,0x60,0xFF,0xFF, 0x80,0x80,0xFF,0xFF,0x80,0x80,0xFF,0xFF,
0xA0,0xA0,0xFF,0xFF,0xA0,0xA0,0xFF,0xFF, 0xC0,0xC0,0xFF,0xFF,0xC0,0xC0,0xFF,0xFF, 0xE0,0xE0,0xFF,0xFF,0xE0,0xE0,0xFF,0xFF, 0x00,0x00,0x21,0x21,0x10,0x10,0xFF,0xFF,
0x23,0x23,0x41,0x41,0x23,0x23,0xFF,0xFF, 0x32,0x32,0x5F,0x5F,0x32,0x32,0xFF,0xFF, 0x41,0x41,0x82,0x82,0x41,0x41,0xFF,0xFF, 0x50,0x50,0xA0,0xA0,0x50,0x50,0xFF,0xFF,
0x5F,0x5F,0xC3,0xC3,0x5F,0x5F,0xFF,0xFF, 0x73,0x73,0xE1,0xE1,0x73,0x73,0xFF,0xFF, 0x85,0x85,0xFF,0xFF,0x7A,0x7A,0xFF,0xFF, 0x91,0x91,0xFF,0xFF,0x6E,0x6E,0xFF,0xFF,
0xA0,0xA0,0xFF,0xFF,0x5F,0x5F,0xFF,0xFF, 0xB4,0xB4,0xFF,0xFF,0x50,0x50,0xFF,0xFF, 0xC3,0xC3,0xFF,0xFF,0x41,0x41,0xFF,0xFF, 0xD2,0xD2,0xFF,0xFF,0x32,0x32,0xFF,0xFF,
0xE1,0xE1,0xFF,0xFF,0x23,0x23,0xFF,0xFF, 0xF0,0xF0,0xFF,0xFF,0x0F,0x0F,0xFF,0xFF, 0x28,0x28,0x28,0x28,0x00,0x00,0xFF,0xFF, 0x40,0x40,0x40,0x40,0x00,0x00,0xFF,0xFF,
0x60,0x60,0x60,0x60,0x00,0x00,0xFF,0xFF, 0x80,0x80,0x80,0x80,0x00,0x00,0xFF,0xFF, 0xA0,0xA0,0xA0,0xA0,0x00,0x00,0xFF,0xFF, 0xC0,0xC0,0xC0,0xC0,0x00,0x00,0xFF,0xFF,
0xE0,0xE0,0xE0,0xE0,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0xFF,0xFF,0x00,0x00,0xFF,0xFF, 0xFF,0xFF,0xFF,0xFF,0x28,0x28,0xFF,0xFF, 0xFF,0xFF,0xFF,0xFF,0x40,0x40,0xFF,0xFF,
0xFF,0xFF,0xFF,0xFF,0x60,0x60,0xFF,0xFF, 0xFF,0xFF,0xFF,0xFF,0x80,0x80,0xFF,0xFF, 0xFF,0xFF,0xFF,0xFF,0xA0,0xA0,0xFF,0xFF, 0xFF,0xFF,0xFF,0xFF,0xC0,0xC0,0xFF,0xFF,
0xFF,0xFF,0xFF,0xFF,0xE0,0xE0,0xFF,0xFF, };

#define fatal(...) do { fprintf(stderr,__VA_ARGS__); fputc('\n',stderr); exit(1); } while(0)

typedef struct {
	unsigned short nimg;
	unsigned short img[128];
} ClassInfo;

static unsigned char head[30];
static int sizeindex,size,rows;
static unsigned char*picture;
static char*classname;
static int classid;
static unsigned char namebuf[512];
static ClassInfo*classes;
static unsigned char*levelpic;

static void
usage(void)
{
	fprintf(stderr,
		"Usage: mb2ff <letter> <?class?> <?index?>                                     \n"
		"                                                                              \n"
		"Extracts pictures from a Hero Mesh puzzle set. mb2ff reads a Free Hero Mesh   \n"
		"composite puzzle set from stdin, extracts pictures and converts them to       \n"
		"farbfeld and writes the result to stdout.                                     \n"
		"                                                                              \n"
		"A Free Hero Mesh composite puzzle set file is a Hamster archive containing the\n"
		"files of a Free Hero Mesh puzzle set, and optionally some additional lumps,   \n"
		"such as FILE_ID.DIZ.                                                          \n"
		"                                                                              \n"
		"This decoder takes up to three arguments:                                     \n"
		"   letter -- a single letter, from A to C, which specifies what size of       \n"
		"             pictures are wanted (this is the only mandatory argument)        \n"
		"   class  -- a class name or a one-based level number. If a class, export all \n"
		"             pictures in a vertical strip. If a level number, render that     \n"
		"             level.                                                           \n"
		"   index  -- a zero-based image number within the class. This is only valid if\n"
		"             the second argument is a class name. It outputs only that one    \n"
		"             picture, rather than all of them.                                \n"
		"                                                                              \n"
		"Pictures are sixteen per row (use ff-strip to change this) and are in the     \n"
		"allocation order; not the order they appear in class definitions.             \n"
		"                                                                              \n"
		"Note when previewing a level, it renders it as it would be in the editor; at  \n"
		"runtime it may look different due to MSG_INIT codes altering stuff.           \n"
		"                                                                              \n"
		"Example usage:                                                                \n"
		"   $ mb2ff A < image.fhm > image.ff                                           \n"
		"   $ mb2ff B 3 < image.fhm > image.ff                                         \n"
		"\n"
	);
	exit(1);
}

static void load_pictures(void) {
	unsigned long n;
	int i,j,k;
	picture=malloc(size*size*rows*16L);
	if(!picture) fatal("Allocation failed");
	// Skip smaller pictures
	for(i=0;i<sizeindex;i++) {
		n=head[i+i+12]|(head[i+i+13]<<8);
		n*=n*rows<<4;
		while(n--) fgetc(stdin);
	}
	// Read requested pictures
	fread(picture,size*rows,size*16L,stdin);
	// Skip larger pictures
	for(i=sizeindex+1;i<head[10];i++) {
		n=head[i+i+12]|(head[i+i+13]<<8);
		n*=n*rows<<4;
		while(n--) fgetc(stdin);
	}
	// Skip picture availability table
	fread(namebuf,1,512,stdin);
	// Skip user message table
	i=fgetc(stdin);
	i|=fgetc(stdin)<<8;
	while(i--) {
		j=fgetc(stdin);
		while(j--) fgetc(stdin);
	}
	// Read classes
	classid=-1;
	n=fgetc(stdin);
	n|=fgetc(stdin)<<8;
	classes=calloc(512,sizeof(ClassInfo));
	if(!classes) fatal("Allocation failed");
	while(n--) {
		// Name
		k=fgetc(stdin);
		fread(namebuf,1,k,stdin);
		namebuf[k]=0;
		// ID number
		j=fgetc(stdin);
		j|=fgetc(stdin)<<8;
		if(classname && !strcasecmp(classname,namebuf)) classid=j;
		if(j>511) fatal("Invalid class number %d",j);
		// Description
		k=fgetc(stdin);
		if(k==255) {
			k=fgetc(stdin);
			k|=fgetc(stdin)<<8;
		}
		while(k--) fgetc(stdin);
		// Attributes
		fread(namebuf,1,62,stdin);
		// Pictures
		k=fgetc(stdin);
		k|=fgetc(stdin)<<8;
		classes[j].nimg=k;
		for(i=0;i<k;i++) {
			fread(namebuf,1,2,stdin);
			if(i<128) classes[j].img[i]=namebuf[0]|(namebuf[1]<<8);
		}
		// User variables
		i=fgetc(stdin);
		i|=fgetc(stdin)<<8;
		while(i--) fread(namebuf,1,8,stdin);
		// Subroutine labels
		i=fgetc(stdin);
		i|=fgetc(stdin)<<8;
		while(i--) fread(namebuf,1,10,stdin);
		// Subroutine codes
		i=fgetc(stdin);
		i|=fgetc(stdin)<<8;
		while(--i) fread(namebuf,1,2,stdin);
		// Message labels
		i=fgetc(stdin);
		i|=fgetc(stdin)<<8;
		while(i--) fread(namebuf,1,10,stdin);
		// Message codes
		for(;;) {
			fread(namebuf,1,2,stdin);
			i=fgetc(stdin);
			i|=fgetc(stdin)<<8;
			if(!i) break;
			i-=2;
			while(i--) fread(namebuf,1,2,stdin);
		}
	}
	if(classname && classid==-1) fatal("No such class: %s",classname);
}

static void load_level(int ok) {
	unsigned long m0=size*29;
	unsigned long m1=size<<4;
	int i,j,n,x,y;
	// Level number
	fread(namebuf,1,2,stdin);
	// Title
	i=fgetc(stdin);
	i|=fgetc(stdin)<<8;
	while(i--) fgetc(stdin);
	// Border colours
	fread(namebuf,1,34,stdin);
	// Objects
	n=fgetc(stdin);
	n|=fgetc(stdin)<<8;
	while(n--) {
		fread(head,1,16,stdin);
		if(ok) {
			unsigned char*p;
			unsigned char*q;
			i=head[0]|(head[1]<<8);
			j=head[2]|(head[3]<<8);
			if(classes[i].nimg<j) j=0;
			x=head[6]-1;
			y=head[8]-1;
			if(x<0 || y<0 || x>29 || y>21) fatal("Coordinates out of range");
			i=classes[i].img[j];
			p=picture+(i&~15)*(size*(unsigned long)size)+(i&15)*(unsigned long)size;
			q=levelpic+(29*y*size+x)*size;
			for(y=0;y<size;y++) for(x=0;x<size;x++) if(p[y*m1+x]) q[y*m0+x]=p[y*m1+x];
		}
	}
	// Level strings
	n=fgetc(stdin);
	n|=fgetc(stdin)<<8;
	while(n--) {
		i=fgetc(stdin);
		i|=fgetc(stdin)<<8;
		while(i--) fgetc(stdin);
	}
}

static void out_picture_1(int id) {
	int x,y;
	unsigned long n=size<<4;
	unsigned char*p=picture+(id&~15)*(size*(unsigned long)size)+(id&15)*(unsigned long)size;
	for(y=0;y<size;y++) for(x=0;x<size;x++) fwrite(palette+(p[y*n+x]<<3),1,8,stdout);
}

static void out_picture_2(int ok,int sz) {
	unsigned long n=rows*((sz<<4)*(unsigned long)sz);
	int c;
	while(n--) {
		c=fgetc(stdin);
		if(ok) fwrite(palette+(c<<3),1,8,stdout);
	}
}

static void out_picture_3(void) {
	unsigned long n=size*29UL*size*21UL;
	unsigned char*p=levelpic;
	while(n--) fwrite(palette+(*p++<<3),1,8,stdout);
}

int main(int argc,char**argv) {
	int i;
	if (argc<2 || !strcmp(argv[1],"-h") || !strcmp(argv[1],"--help"))
		usage();
	fread(head,1,30,stdin);
	if(memcmp(head+2,"MESH",4)) fatal("Improper file format");
	if(argv[1][0]>='A' && argv[1][0]<head[10]+'A') sizeindex=argv[1][0]-'A';
	if(argv[1][0]>='a' && argv[1][0]<head[10]+'a') sizeindex=argv[1][0]-'a';
	else fatal("Invalid picture size");
	size=head[sizeindex+sizeindex+12]|(head[sizeindex+sizeindex+13]<<8);
	rows=head[28]|(head[29]<<8);
	for(i=0;i<rows;i++) fgetc(stdin),fgetc(stdin);
	if(argc==2) {
		// Dump all pictures
		fwrite("farbfeld",1,8,stdout);
		putchar(0); putchar(size>>12); putchar(size>>4); putchar(size<<4);
		i=size*rows;
		putchar(i>>24); putchar(i>>16); putchar(i>>8); putchar(i);
		for(i=0;i<sizeindex;i++) out_picture_2(0,head[i+i+12]|(head[i+i+13]<<8));
		out_picture_2(1,size);
	} else if(argc==3 && ((argv[2][0]>='0' && argv[2][0]<='9') || argv[2][0]=='-')) {
		// Preview a level
		load_pictures();
		levelpic=calloc(29*size,21*size);
		if(!levelpic) fatal("Allocation failed");
		i=strtol(argv[2],0,0);
		fread(namebuf,1,4,stdin);
		if(i<1 || i>(namebuf[0]|(namebuf[1]<<8))) fatal("Level number out of range");
		while(i--) load_level(!i);
		fwrite("farbfeld",1,8,stdout);
		i=size*29;
		putchar(i>>24); putchar(i>>16); putchar(i>>8); putchar(i);
		i=size*21;
		putchar(i>>24); putchar(i>>16); putchar(i>>8); putchar(i);
		out_picture_3();
	} else if(argc==3) {
		// All pictures of a class
		classname=argv[2];
		load_pictures();
		fwrite("farbfeld",1,8,stdout);
		putchar(size>>24); putchar(size>>16); putchar(size>>8); putchar(size);
		i=size*classes[classid].nimg;
		putchar(i>>24); putchar(i>>16); putchar(i>>8); putchar(i);
		for(i=0;i<classes[classid].nimg;i++) out_picture_1(classes[classid].img[i]);
	} else if(argc==4) {
		// One picture of a class
		classname=argv[2];
		load_pictures();
		i=strtol(argv[3],0,0);
		if(i<0 || i>=classes[classid].nimg) fatal("Image number out of range; valid range is 0 to %d",classes[classid].nimg-1);
		fwrite("farbfeld",1,8,stdout);
		putchar(size>>24); putchar(size>>16); putchar(size>>8); putchar(size);
		putchar(size>>24); putchar(size>>16); putchar(size>>8); putchar(size);
		out_picture_1(classes[classid].img[i]);
	} else {
		fatal("Incorrect number of arguments");
	}
	return 0;
}
